ang/civicase/case/actions/services/go-to-webform-case-action.service.js

```diff
@@ -0,0 +1,26 @@
+(function (angular, $, _) {
+  var module = angular.module('civicase');
+
+  module.service('GoToWebformCaseAction', GoToWebformCaseAction);
+
+  function GoToWebformCaseAction ($window) {
+    /**
+     * Click event handler for the Action
+     *
+     * @param {Array} cases
+     * @param {Object} action
+     * @param {Function} callbackFn
+     */
+    this.doAction = function (cases, action, callbackFn) {
+      var window;
+      var urlObject = { case1: cases[0].id };
+
+      if (action.clientID) {
+        urlObject['cid' + action.clientID] = cases[0].client[0].contact_id;
+      }
+
+      window = $window.open(CRM.url(action.path, urlObject), '_blank');
+      window.focus();
+    };
```

Comment:
sometimes we return an url, sometimes we do an actual action. Can't the ones that return an URL do an action instead and move all the logic for opening the CRM form into a service that these actions can use?

Reply:
I agree, it can be refactored that way. But given the timeline, I dont want to increase the scope of this ticket even more. For now only the existing code is moved inside Services, without any kind of refactoring.

Reply:
Added a TODO